### ISSUE TITLE
Building with crt-builder fails on OpenBSD

### ISSUE THREAD

**Issue by knightjoel** [awslabs/aws-c-io#548]
Hi. With the merging of #545 and awslabs/aws-crt-builder#217, I've been working on adding a CI job for OpenBSD to aws-c-io. When I build aws-c-io using the cmake commands in README.md, everything builds and tests just fine. When I use crt-builder, it fails to link:

```
% aws-crt-builder/build/builder build -p aws-c-io
[...]
[100%] Linking C executable aws-c-io-tests
ld: error: undefined symbol: CRYPTO_get_id_callback
>>> referenced by openssl_platform_init.c:565 (/home/joel/aws-c-io/build/deps/aws-c-cal/source/unix/openssl_platform_init.c:565)
>>>               openssl_platform_init.c.o:(aws_cal_platform_init) in archive /home/joel/aws-c-io/build/install/lib/libaws-c-cal.a
>>> referenced by openssl_platform_init.c:582 (/home/joel/aws-c-io/build/deps/aws-c-cal/source/unix/openssl_platform_init.c:582)
>>>               openssl_platform_init.c.o:(aws_cal_platform_clean_up) in archive /home/joel/aws-c-io/build/install/lib/libaws-c-cal.a
>>> did you mean: CRYPTO_set_id_callback
>>> defined in: /home/joel/aws-c-io/build/install/lib/libcrypto.a(thread.c.o)
clang: error: linker command failed with exit code 1 (use -v to see invocation)
*** Error 1 in . (tests/CMakeFiles/aws-c-io-tests.dir/build.make:439 'tests/aws-c-io-tests')
*** Error 2 in . (CMakeFiles/Makefile2:882 'tests/CMakeFiles/aws-c-io-tests.dir/all')
*** Error 2 in /home/joel/aws-c-io/build/aws-c-io (Makefile:146 'all': /usr/bin/make -s -f CMakeFiles/Makefile2 all)
Failed to run cmake --build /home/joel/aws-c-io/build/aws-c-io --config RelWithDebInfo --target install: Command exited with code 2
```

Note the error is bubbling up from aws-c-cal, however I can  run `builder build -p aws-c-cal` directly and it builds and tests just fine on its own.

Looking at `aws-c-io/build/deps/aws-c-cal/source/unix/openssl_platform_init.c`, it looks like `OPENSSL_IS_AWSLC` is not being defined when building via crt-builder. I'm running in circles trying to debug what conditions cause that define to be set and why things are different under crt-builder vs running cmake directly. I'm looking for insight or pointers that will unblock me. Thanks for reading.

**Comment by graebm** [comment]
Huh. Seems like the build is maybe getting confused between your system's OpenSSL installation and AWS-LC. Like, maybe it's using the OpenSSL headers, but the AWS-LC static library.

Can you paste the cmake configuration line? where it's passing `-DCMAKE_PREFIX_PATH` and such?

**Comment by graebm** [comment]
just post the full build logs. Attach it as a file if it's too long for a normal comment box

**Comment by knightjoel** [comment]
Build log attached. You'll need this diff to aws-c-io for the build to get all the way to linking.

```diff
--- a/builder.json
+++ b/builder.json
@@ -5,8 +5,8 @@
         { "name": "aws-c-cal" },
         {
             "name": "s2n",
-            "revision": "v1.3.11",
-            "targets": ["linux", "android"]
+            "revision": "v1.3.36",
+            "targets": ["linux", "android", "openbsd"]
         }
     ],
     "downstream": [

```

[build.log](https://github.com/awslabs/aws-c-io/files/10743816/build.log)

**Comment by bretambrose** [comment]
Perhaps an analogous change to aws-c-cal is necessary (if not already done locally)?  I see entries for linux and android mapping to aws-lc in its builder.json

**Comment by knightjoel** [comment]
Yet, aws-c-cal builds just fine if I do a standalone build of it (`aws-crt-builder/build/builder build -p aws-c-cal`).

**Comment by bretambrose** [comment]
Sure, but is your standalone build of aws-c-cal linking against openssl or aws-lc?

**Comment by knightjoel** [comment]
Here's what I see building aws-c-cal on its own using builder. Is this what I should be looking for to see where it's pulling libcrypto from?

```
> cmake -B/home/jwk/aws-c-cal/build/aws-c-cal -H/home/jwk/aws-c-cal -DAWS_WARNINGS_ARE_ERRORS=ON -DCMAKE_INSTALL_PREFIX=/home/jwk/aws-c-cal/build/install -DCMAKE_PREFIX_PATH=/hom
e/jwk/aws-c-cal/build/install -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_BUILD_TYPE=RelWithDebInfo -DBUILD_TESTING=ON -DCMAKE_C_COMPILER=/usr/bin/clang -DCMAKE_CXX_COMPILER=/usr/
bin/clang++ -DENABLE_SANITIZERS=OFF
-- The C compiler identification is Clang 13.0.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/clang - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Found crypto: /usr/lib/libcrypto.a
-- LibCrypto Include Dir: /usr/include
-- LibCrypto Shared Lib:  crypto_SHARED_LIBRARY-NOTFOUND
-- LibCrypto Static Lib:  /usr/lib/libcrypto.a
```

Let me tune builder.json in aws-c-cal and run a build again.

**Comment by knightjoel** [comment]
Well spotted, @bretambrose. With the following diff to aws-c-cal, builder builds and runs tests for aws-c-io.

```diff
--- a/builder.json
+++ b/builder.json
@@ -25,6 +25,13 @@
                     "name": "aws-lc"
                 }
             ]
+        },
+        "openbsd": {
+            "upstream": [
+                {
+                    "name": "aws-lc"
+                }
+            ]
         }
     },
     "variants": {
```

The aws-c-io test suite fails multiple TLS tests, but that's a different problem. Thanks for the pointers, both.
